### PR TITLE
Add QR code generation and scanning to UI

### LINKED ISSUE #803: QR codes and barcodes for samples
cc @PeterKraus @jdbocarsly (and maybe @MJCliffe?)

We have often discussed labelling physical samples with QR codes/bar codes. This issue can be used to outline the work needed and then potential extensions.

- [ ] Resolve sample pages by refcode in the  API and UI

then either:

- [ ] We add a configurable "labelling" block that allows you to choose QR code size, style etc. for printing, and any extra info you want to be on the label
- [ ] or, we hardcode the label style at the level of the server, and always show that label on the item page. How it prints at different scales will be up to the user.

# Open questions

- How should permissions work in this case? We could embed a read-only token for the QR code, so if someone can scan your samples, then they should at least be able to see what it is/was.

# Extensions

## Link resolvers

How long do these barcodes need to last, and what info should we try to encode in them in case the datalab is gone/dead? I think there are two options:

- [ ] Stuff as much of the JSON data into the QR code itself, e.g., formula, hazards and so on, so that it at least can be resolved by a smart QR code scanner if datalab is dead. I don't know how smart a normal mobile browser scanner is, so this would need some experimenting.
- [ ] Devise a "datalab federation" link resolver, so that the QR code only points to e.g., `datalab.ly/concat:refcode`, then a simple server at datalab.ly resolves that `concat:` currently exists at `datalab.concatlab.eu`. This could allow us to have an interesting "off-ramp" where we statically host an entire datalab's JSON response to allow this resolution to work "forever".

## Printing labels from datalab

- [ ] Attach label printer to the datalab server itself (via tunnels or whatever), so that a "print label" button always goes to a printer in the lab, with some predefined sizes.
- [ ] I think @jdbocarsly was interested in the reverse, where the act of printing a label actually creates the stub datalab page and reserves the ID.

## Barcode scanning in datalab itself

- [ ] Given the read-only permissions above, it would be nice to use QR codes to also ingest data into datalab. e.g., you exchange samples with a collaborator, they can scan the QR code via the datalab browser (either uploading the image, or using the camera functionality), and then datalab can make the API call to the other person's datalab and ingest it, maintaining various backlinks. In this case, it could be very nice to even communicate to the "remote" datalab that the sample has been "received" and that it is now locked for editing until they rescan the QR code at original end (or alternatively, until they rescan it, any changes made remotely must be pushed back to the local datalab).

### PR DESCRIPTION
**Pull request by ml-evs**
Closes #803.

This PR:

- [x] Adds a QR code icon to all refcode displays, which, when clicked, opens a modal showing the QR code for that sample item, with a WIP print button that create a dummy window containing just the QR code.
- [x] The QR codes point to `{{ QR_CODE_RESOLVER_URL }}/items/<refcode>`, which can be configured. The refcode itself is always parseable from this URL, so datalab-native scanners (such as that described below) can resolve the item without needing to go via the resolver.
- [x] There is also a "Scan QR Code" button at the top of the sample table page (for now) which opens a camera stream and provides file upload for providing a datalab QR code.
- [x] API route argument is added for `redirect-to-ui`, where the `/items` endpoint can point itself to the UI, if configured. This is the default redirect used by the purl system.

Potential follow-ups:

- [ ] This scanner should do the following:
    - [ ] Detect if the sample is from this deployment, and if so, redirect to the edit page directly.
    - [ ] Follow the resolver link with JS and get the "real" underlying URL, then present it to the user to click on, if it believes it to be a datalab QR code.
    - [ ]  If it does not believe it to be a datalab QR code, do the same as above but with a giant warning.
- [ ] QR codes should also embed tokens that allow for access of non-private/safety data. This will need updates to the API too.
- [ ] The barcode print itself should also allow some info to be printed automatically.

### CODE REVIEW

File: webapp/src/resources.js [hunk 1]
```diff
@@ -24,6 +24,11 @@ export const API_URL =
   process.env.VUE_APP_API_URL != null ? process.env.VUE_APP_API_URL : "http://localhost:5001";
 export const API_TOKEN = process.env.VUE_APP_API_TOKEN;
 
+export const QR_CODE_RESOLVER_URL =
+  process.env.VUE_APP_QR_CODE_RESOLVER_URL != null
+    ? process.env.VUE_APP_QR_CODE_RESOLVER_URL
+    : "https://purl.datalab-org.io";
```
Comment (ml-evs) [inline webapp/src/resources.js:30]: I think this is the bit I'm still unsure about; think we need another option for whether the link resolver should be turned on; or even a server value for "is this server in the federation", and if not, just use the datalab's URL as the resolver

**Review by ml-evs — COMMENTED** [review]
This this is ready when you are @jdbocarsly and @BenjaminCharmes -- I have one comment we can discuss regarding configuration (but probably that can be addressed in another PR before release)

**Review by BenjaminCharmes — APPROVED** [review]
It looks really good! 👏

**Review by jdbocarsly — APPROVED** [review]
looks good! excited to start using this and build on top of it

**Pull request merged**